lib/rust/ensogl/component/grid-view/src/header.rs

```diff
@@ -0,0 +1,655 @@
+        let cols_range = visible_area::visible_columns(viewport, entries_size, col_count);
+        let highest_visible_row =
+            visible_area::visible_rows(viewport, entries_size, row_count).start;
+        let freed = visible_headers.drain_filter(|col, header| {
+            !cols_range.contains(col) || !header.section_rows.contains(&highest_visible_row)
+        });
+        let detached = freed.map(|(col, header)| {
+            header.entry.entry.unset_parent();
+            ((header.section_rows.start, col), header.entry)
+        });
+        let (locations, entries): (Vec<_>, Vec<_>) = detached.unzip();
+        free_headers.extend(entries);
+        locations
+    }
+
+    fn needed_info_for_uncovered_sections(&self, properties: Properties) -> Vec<(Row, Col)> {
+        let Properties { row_count, col_count, viewport, entries_size } = properties;
+        let visible_headers = self.visible_headers.borrow();
+        let cols_range = visible_area::visible_columns(viewport, entries_size, col_count);
+        let highest_visible_row =
+            visible_area::visible_rows(viewport, entries_size, row_count).start;
+        let uncovered = cols_range.filter_map(|col| {
+            (!visible_headers.contains_key(&col)).as_some((highest_visible_row, col))
+        });
+        uncovered.collect()
+    }
+
+    fn reset_entries(&self, properties: Properties) -> Vec<(Row, Col)> {
+        let Properties { row_count, col_count, viewport, entries_size } = properties;
+        let mut visible_headers = self.visible_headers.borrow_mut();
+        let mut free_headers = self.free_headers.borrow_mut();
+        let highest_visible_row =
+            visible_area::visible_rows(viewport, entries_size, row_count).start;
+        let detached = visible_headers.drain().map(|(_, header)| {
+            header.entry.entry.unset_parent();
+            header.entry
+        });
+        free_headers.extend(detached);
+        let visible_columns = visible_area::visible_columns(viewport, entries_size, col_count);
+        visible_columns.map(|col| (highest_visible_row, col)).collect()
+    }
+
+    fn drop_all_entries(&self, properties: Properties) -> Vec<(Row, Col)> {
+        let to_section_info_request = self.reset_entries(properties);
+        self.free_headers.borrow_mut().clear();
+        to_section_info_request
+    }
+}
+
+impl<InnerGrid, HeaderEntry: Entry> Model<InnerGrid, HeaderEntry, HeaderEntry::Params> {
+    fn header_position(
+        &self,
+        row: Row,
+        col: Col,
+        entry_size: Vector2,
+        viewport: Viewport,
+    ) -> Option<Vector2> {
+        let visible_headers = self.visible_headers.borrow();
+        let header = visible_headers.get(&col).filter(|h| h.section_rows.start == row);
+        header.map(|h| h.header_position(col, entry_size, viewport))
+    }
+
+    fn header_separator(&self, col: Col, entry_size: Vector2, viewport: Viewport) -> f32 {
```

Comment:
Some doc for the return value might be nice. Is it the y offset from some baseline?